src/ListMessageSplitter.ts

```diff
@@ -0,0 +1,319 @@
+    //
+    // Returns:
+    // `sized` != null, if sized was adequate, or had to be split
+    // `rest` != null, if sized was not adequate (and had to be split),
+    //   or the first item is too big to be split at the desired size.
+    public split_at_size(desired_size: number): {sized: MessageListing | null, rest: MessageListing | null} {
+        if (this.size() <= desired_size) {
+            // If the current listing is under the desired size, just return it, rest = null.
+            return {sized: this, rest: null};
+        } else {
+            // Else, split it.
+
+            // Create a new listing with just the current header,
+            // as a candidate for the new sized listing.
+            let sized = new MessageListing(this.header);
+            if (sized.size() > desired_size) {
+                // If the header alone is too much, just give up.
+                return {sized: null, rest: this}
+            }
+
+            // Create another listing, and dump the rest of all items in there.
+            let rest = new MessageListing(null);
+            // Be sure to shallow copy, to not disturb the `this` listing.
+            rest.items = [...this.items];
+
+            // Perform initial shift of item into a variable.
+            let current_item = rest.items.shift();
+
+            // Keep looping while there are still items left.
+            // (Replaced at the end of the loop)
+            while (current_item !== undefined) {
+                // Add the new item to `sized` tentatively.
+                sized.items.push(current_item);
+
+                if (sized.size() > desired_size) {
+                    // If we went over the limit, return the last item and return the results
+                    rest.items.unshift(sized.items.pop()!)
+
+                    if (sized.items.length > 0) {
+                        return {sized, rest};
+                    } else {
+                        // If this was the first item, `sized` is empty, return null.
+                        return {sized: null, rest}
+                    }
+                }
+
+                current_item = rest.items.shift();
+            }
+
+            // ??? We somehow did not go over the size limit when testing it per item?
+            // This is a weird state, as we should have already caught this with the
+            // `this.size() <= desired_size` conditional.
+            LogService.warn("ListMessageSplitter", `encountered end of while loop, required max ${desired_size} size, current item is ${sized.size()}`)
+
+            return {sized, rest: null}
+        }
+    }
+
+    public size(): number {
+        let with_header = this.header !== null ? this.header.size() + "\n".length + "<br>".length : 0;
+
+        return with_header
+            + (this.items.length > 0 ? UL_TAG_WRAP_SIZE : 0)
+            + this.items.reduce((prev, curr, _idx, _arr) => prev + curr.size(), 0);
```

Comment:
btw this can just be written as `(prev, curr)` btw, you don't need to accept arguments you're not using